### ISSUE TITLE
Support for mocking

### ISSUE THREAD

**Issue by bhurlow** [cognitect-labs/aws-api#186]
Our team is looking for a way to mock certain operations invoked by `aws-api` during unit tests and certain local development contexts. Is there a preferred mocking strategy with this library? 

Initially I thought that `cognitect.aws.client/send-request` was defined as a protocol and thus could be overridden by creating a mock client, but it appears that it is called directly. For now, we've been able to use `with-redefs` to short circuit the sending of the request, but this adds additional test overhead. 

Would the authors be open to adjusting this pattern to support mocking use cases? 

Thanks!

**Comment by dchelimsky** [comment]
We're working on something that looks like this right now:

```clojure
(require '[cognitect.aws.client.test :as test]
         '[cognitect.aws.client.api :as aws])

;; implementation being tested
(defn do-something [s3-client]
  ;; if :response is bound to a value, invoke will return that value
  ;; if :response is bound to a funtion, invoke will return the return
  ;;    of that function
  (let [res (aws/invoke s3-client {:op :CreateBucket
                                   :request {:Bucket "a-bucket"}})]
    ;; do stuff with res
    ))

;; test using a stub
(let [test-s3-client (test/client {:api :s3})]
  (test/stub test-s3-client
             {:op :ListBuckets
              :response {:Location "a-location"}})
  (let [res (do-something test-s3-client)]
    ;; make assertions about res
    ))

;; test using a handler function
(let [test-s3-client (test/client {:api :s3})]
  (test/stub test-s3-client
             {:op :ListBuckets
              ;; op-map is the 2nd argument to `invoke`
              :response (fn [op-map]
                          ;; do stuff based on op-map
                          ;; then, return a value
                          {:Location "def"})})
  (let [res (do-something test-s3-client)]
    ;; make assertions about res
    ))
```

In most cases, we’d expect that folks would just stub return values, but for those who want more interaction based tests (mocks, spies, etc), the functions let you implement whatever you want.

Also, this is based on a new Client protocol with invoke and stop, that you can reify yourself and do whatever you want with.

Thoughts? This is in progress now and subject to change, so any feedback would be most appreciated.

**Comment by bsless** [comment]
The response function looks good

**Comment by dchelimsky** [comment]
We've got an implementation of this on the `mcrmc/test-client` branch and would like some feedback. To try it out:

```shell
git clone git@github.com:cognitect-labs/aws-api.git
git co mcrmc/test-client
```

Now you can include it in a local project with `{:deps {com.cognitect.aws/api {:local/root "<path to aws-api directory>"}}}`

Feedback please!

**Comment by dchelimsky** [comment]
Released in version 0.8.589
